src/main/java/net/wasdev/wlp/common/plugins/util/InstallFeatureUtil.java

```diff
@@ -0,0 +1,683 @@
+     * @return whether debug is enabled
+     */
+    public abstract boolean isDebugEnabled();
+
+    /**
+     * Download the artifact from the specified Maven coordinates, or retrieve it from the cache if it already exists.
+     * 
+     * @param groupId The group ID
+     * @param artifactId The artifact ID
+     * @param type The type e.g. esa
+     * @param version The version
+     * @return The file corresponding to the downloaded artifact
+     * @throws PluginExecutionException If the artifact could not be downloaded
+     */
+    public abstract File downloadArtifact(String groupId, String artifactId, String type, String version)
+            throws PluginExecutionException;
+    
+    /**
+     * Combine the given String collections into a set
+     * 
+     * @param collections a collection of strings
+     * @return the combined set of strings
+     */
+    @SafeVarargs
+    public static Set<String> combineToSet(Collection<String>... collections) {
+        Set<String> result = new HashSet<String>();
+        for (Collection<String> collection : collections) {
+            result.addAll(collection);
+        }
+        return result;
+    }
+    
+    /**
+     * Get the set of features defined in the server.xml if there were no features listed in the plugin configuration
+     * @param serverDirectory The server directory containing the server.xml
+     * @param noPluginListedFeatures true if there were no features listed in the plugin configuration
+     * @return the set of features that should be installed from server.xml
+     */
+    public Set<String> getServerFeatures(File serverDirectory, boolean noPluginListedFeatures) {
+        // parse server.xml features only if there are no configured features in the pom
+        if (noPluginListedFeatures) {
+            debug("No features were listed for the plugin. Using server.xml.");
+            return getServerFeatures(serverDirectory);
+        } else {
+            debug("Features were listed for the plugin. Skipping server.xml.");
+            return new HashSet<String>();
+        }
+    }
+
+    /**
+     * Get the set of features defined in the server.xml
+     * @param serverDirectory The server directory containing the server.xml
+     * @return the set of features that should be installed from server.xml
+     */
+    public static Set<String> getServerFeatures(File serverDirectory) {
+        Set<String> result = new HashSet<String>();
+        result.addAll(getConfigDropinFeatures(serverDirectory, "overrides"));
+        result.addAll(getServerXmlFeatures(new File(serverDirectory, "server.xml"), null));
+        result.addAll(getConfigDropinFeatures(serverDirectory, "defaults"));
+        return result;
+    }
+    
+    private static Set<String> getConfigDropinFeatures(File serverDirectory, String folderName) {
+        Set<String> result = new HashSet<String>();
+        String configDropins = serverDirectory.getAbsolutePath() + "/configDropins/" + folderName;
```

Comment:
Replace `getAbsolutePath()` with `getCanonicalPath()` throughout. For example, `/a/b/c/../` is an absolute path while `/a/b/` is the equivalent canonical path. If you are doing file comparisons you will want to compare the canonical files.

Reply:
Changed to use getCanonicalFile throughout where appropriate.